add: post on deploying tf vision models with tf serving. (#427)
huggingface/blog: tf-serving-vision.md

        <img class="avatar avatar-user" src="https://avatars.githubusercontent.com/u/22957388?v=4" title="Gravatar">
        <div class="bfc">
            <code>sayakpaul</code>
            <span class="fullname">Sayak Paul</span>
            <span class="bg-gray-100 dark:bg-gray-700 rounded px-1 text-gray-600 text-sm font-mono">guest</span>
        </div>
    </a>
</div>

<a target="_blank" href="https://colab.research.google.com/github/sayakpaul/deploy-hf-tf-vision-models/blob/main/hf_vision_model_tfserving.ipynb">
    <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

In the past few months, the Hugging Face team and external contributors
added a variety of vision models in TensorFlow to Transformers. This
list is growing comprehensively and already includes state-of-the-art
pre-trained models like [Vision Transformer](https://huggingface.co/docs/transformers/main/en/model_doc/vit),
[Masked Autoencoders](https://huggingface.co/docs/transformers/model_doc/vit_mae),
[RegNet](https://huggingface.co/docs/transformers/main/en/model_doc/regnet),
[ConvNeXt](https://huggingface.co/docs/transformers/model_doc/convnext),

> and so on.

Author: There are more vision models in TF, which I didn't mention for brevity. What would you suggest in that case?

> and so on.

Reviewer: maybe ", and many others!" WDYT?

> and so on.

Author: Yes, sure!

When it comes to deploying TensorFlow models, we've got a variety of
options. Depending on your use case, you may want to expose your model
as an endpoint or package it in an application itself. TensorFlow
provides tools that cater to each of these different scenarios.

In this post, we'll see how to deploy a Vision Transformer (ViT) model
locally using [TensorFlow Serving](https://www.tensorflow.org/tfx/tutorials/serving/rest_simple)
(TF Serving). This will allow developers to expose the model either as a
REST or gRPC endpoint. Moreover, TF Serving supports many
deployment-specific features off-the-shelf such as model warmup,
server-side batching, etc.

To get the complete working code shown throughout this post, refer to
the Colab Notebook shown at the beginning.

# Saving the Model

All TensorFlow models in 🤗 Transformers have a method named
`save_pretrained()`. With it, we can serialize the model weights in
the h5 format as well as in the standalone [SavedModel format](https://www.tensorflow.org/guide/saved_model).
TF Serving needs a model to be present in the SavedModel format. So, let's first
load a Vision Transformer model and save it:

> 

Author: Sure.

```py
from transformers import TFViTForImageClassification

temp_model_dir = "vit"
ckpt = "google/vit-base-patch16-224"

model = TFViTForImageClassification.from_pretrained(ckpt)
model.save_pretrained(temp_model_dir, saved_model=True)
```

By default, `save_pretrained()` will first create a version directory
inside the path we provide to it. So, the path ultimately becomes:
`{temp_model_dir}/saved_model/{version}`.

We can inspect the serving signature of the SavedModel like so:

```bash
saved_model_cli show --dir {temp_model_dir}/saved_model/1 --tag_set serve --signature_def serving_default
```

This should output:

```bash
The given SavedModel SignatureDef contains the following input(s):
  inputs['pixel_values'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, -1, -1, -1)
      name: serving_default_pixel_values:0
The given SavedModel SignatureDef contains the following output(s):
  outputs['logits'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 1000)
      name: StatefulPartitionedCall:0
Method name is: tensorflow/serving/predict
```

> As can be noticed the model accepts single 4-d inputs (namely

Author: I intentionally kept it to `notice` for being more inclusive as not everyone can see :) But okay to change to `see` if that's preferred.

> As can be noticed the model accepts single 4-d inputs (namely

Reviewer: Actually you're totally right, my fault! Let's keep the noticed!

- `model_name` specifies the model name (can be anything) that will
  used for calling the APIs.

- `model_base_path` denotes the base model path that TF Serving will
  use to load the latest version of the model.

(The complete list of supported parameters is
[here](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/model_servers/main.cc).)

And voila! Within minutes, you should be up and running with a deployed
model having two endpoints - REST and gRPC.

# Querying the REST Endpoint

Recall that we exported our model such that it accepts string inputs
encoded with the [base64 format](https://en.wikipedia.org/wiki/Base64). So, to craft our
request payload we'll do something like this:

```py
# Get image of a cute cat.
image_path = tf.keras.utils.get_file(
    "image.jpg", "http://images.cocodataset.org/val2017/000000039769.jpg"
)

# Read the image from disk as raw bytes and then encode it. 
bytes_inputs = tf.io.read_file(image_path)
b64str = base64.urlsafe_b64encode(bytes_inputs.numpy()).decode("utf-8")

# Create the request payload.
data = json.dumps({"signature_name": "serving_default", "instances": [b64str]})
```

> TF Serving's request payload format specification for the REST endpoint
> is available [here](https://www.tensorflow.org/tfx/serving/api_rest#request_format_2).
> Within the `instances` we can pass multiple encoded images. This kind
> of endpoints are meant to be consumed for online prediction scenarios.
> For inputs having more than a single data point, you would to want to
> [enable batching](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/batching/README.md)
> to get performance optimization benefits.
> 
> Now we can call the API:
> 

Reviewer: ```suggestion
TF Serving's request payload format specification for the REST endpoint
is available [here](https://www.tensorflow.org/tfx/serving/api_rest#request_format_2).
Within the `instances`, you can pass multiple encoded images. This kind
of endpoints are meant to be consumed for online prediction scenarios.
For inputs having more than a single data point, you would to want to
[enable batching](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/batching/README.md)
to get performance optimization benefits.

Now you can call the API:

```

```py
headers = {"content-type": "application/json"}
json_response = requests.post(
    "http://localhost:8501/v1/models/vit:predict", data=data, headers=headers
)
print(json.loads(json_response.text))
# {'predictions': [{'label': 'Egyptian cat', 'confidence': 0.896659195}]}

```

> Our REST API is -
> `http://localhost:8501/v1/models/vit:predict` following the specification from
> [here](https://www.tensorflow.org/tfx/serving/api_rest#predict_api). By default,
> this always picks up the latest version of the model. But if we wanted a
> specific version we can do: `http://localhost:8501/v1/models/vit/versions/1:predict`.

Reviewer: ```suggestion
The REST API is -
`http://localhost:8501/v1/models/vit:predict` following the specification from
[here](https://www.tensorflow.org/tfx/serving/api_rest#predict_api). By default,
this always picks up the latest version of the model. But if you wanted a
specific version you can use: `http://localhost:8501/v1/models/vit/versions/1:predict`.
```

# Querying the gRPC Endpoint

> While REST is quite popular in the API world many applications often

Reviewer: ```suggestion
While REST is quite popular in the API world, many applications often
```

benefit from gRPC. [This post](https://blog.dreamfactory.com/grpc-vs-rest-how-does-grpc-compare-with-traditional-rest-apis/)
does a good job comparing the two ways of deployment. gRPC is usually
preferred for low-latency, highly scalable, and distributed systems.

> There are a couple of steps are. First, we need to open a communication

Reviewer: ```suggestion
There are a couple of steps are. First, you need to open a communication
```

channel:

```py
import grpc
from tensorflow_serving.apis import predict_pb2
from tensorflow_serving.apis import prediction_service_pb2_grpc

channel = grpc.insecure_channel("localhost:8500")
stub = prediction_service_pb2_grpc.PredictionServiceStub(channel)
```

> Then, we create the request payload:

Reviewer: ```suggestion
Then, create the request payload:
```

```py
request = predict_pb2.PredictRequest()
request.model_spec.name = "vit"
request.model_spec.signature_name = "serving_default"
request.inputs[serving_input].CopyFrom(tf.make_tensor_proto([b64str]))
```

> We can determine the `serving_input` key programmatically like so:

Reviewer: ```suggestion
You can determine the `serving_input` key programmatically like so:
```

```py
loaded = tf.saved_model.load(f"{MODEL_DIR}/{VERSION}")
serving_input = list(
    loaded.signatures["serving_default"].structured_input_signature[1].keys()
)[0]
print("Serving function input:", serving_input)
# Serving function input: string_input
```

> Now, we can get some predictions:

Reviewer: ```suggestion
Now, you can get some predictions:
```

```py
grpc_predictions = stub.Predict(request, 10.0)  # 10 secs timeout
print(grpc_predictions)
```

```bash
outputs {
  key: "confidence"
  value {
    dtype: DT_FLOAT
    tensor_shape {
      dim {
        size: 1
      }
    }
    float_val: 0.8966591954231262
  }
}
outputs {
  key: "label"
  value {
    dtype: DT_STRING
    tensor_shape {
      dim {
        size: 1
      }
    }
    string_val: "Egyptian cat"
  }
}
model_spec {
  name: "resnet"
  version {
    value: 1
  }
  signature_name: "serving_default"
}
```

> We can also fetch the key-values of our interest from the above results like so:

Reviewer: ```suggestion
You can also fetch the key-values of our interest from the above results like so:
```

```py
grpc_predictions.outputs["label"].string_val, grpc_predictions.outputs[
    "confidence"
].float_val
# ([b'Egyptian cat'], [0.8966591954231262])
```

# Wrapping Up

> In this post, we learned how to deploy a TensorFlow vision model from
> Transformers with TF Serving. While local deployments are great for
> weekend projects, we would want to be able to scale these deployments to
> serve many users. In the next series of posts, we'll see how to scale up
> these deployments with Kubernetes and Vertex AI.

Reviewer: ```suggestion
In this post, we learned how to deploy a TensorFlow vision model from
Transformers with TF Serving. While local deployments are great for
weekend projects, we would want to be able to scale these deployments to
serve many users. In the next series of posts, you'll see how to scale up
these deployments with Kubernetes and Vertex AI.
```

---

> title: Deploying TensorFlow Vision Models in Hugging Face with TF Serving

Author: Not sure how to generate a thumbnail. Any resource will be helpful. I can add it after that.

Author: @Reeray I am happy to follow any instructions too. Appreciate the help in advance.

---

<h1>
  Deploying TensorFlow Vision Models in Hugging Face with TF Serving
</h1>

<div class="blog-metadata">

>     <small>Published July 19, 2022.</small>

Author: Of course, will change this to the merge date after the PR is approved?
